gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/HiveMetadataWriter.java

```diff
@@ -0,0 +1,242 @@
+      this.hiveRegister.createTableIfNotExists(tableSpec.getTable());
+    }
+    if (!lastestSchemaMap.containsKey(tableKey)) {
+      HiveTable existingTable = this.hiveRegister.getTable(dbName, tableName).get();
+      lastestSchemaMap.put(tableKey, existingTable.getSerDeProps().getProp(
+          AvroSerdeUtils.AvroTableProperties.SCHEMA_LITERAL.getPropName()));
+    }
+
+    //Calculate the topic name from gmce, fall back to topic.name in hive spec which can also be null
+    //todo: make topicName fall back to topic.name in hive spec so that we can also get schema for re-write operation
+    String topicName = null;
+    if (gmce.getTopicPartitionOffsetsRange() != null) {
+      String topicPartitionString = gmce.getTopicPartitionOffsetsRange().keySet().iterator().next();
+      //In case the topic name is not the table name or the topic name contains '-'
+      topicName = topicPartitionString.substring(0, topicPartitionString.lastIndexOf('-'));
+    }
+    switch (gmce.getOperationType()) {
+      case add_files: {
+        addFiles(gmce, newSpecsMap, dbName, tableName, topicName);
+        break;
+      }
+      default: {
+        log.error("unsupported operation {}", gmce.getOperationType().toString());
+        return;
+      }
+    }
+  }
+
+  public void addFiles(GobblinMetadataChangeEvent gmce, Map<String, Collection<HiveSpec>> newSpecsMap, String dbName, String tableName, String topicName)
+      throws SchemaRegistryException {
+    String tableKey = tableNameJoiner.join(dbName, tableName);
+    for (Collection<HiveSpec> specs : newSpecsMap.values()) {
+      for (HiveSpec spec: specs) {
+         if (spec.getTable().getDbName().equals(dbName) && spec.getTable().getTableName().equals(tableName)) {
+           List<String> partitionValue = spec.getPartition().isPresent()? spec.getPartition().get().getValues() : Lists.newArrayList();
+           Cache<List<String>, HiveSpec> hiveSpecCache = specMaps.computeIfAbsent(tableKey, s -> CacheBuilder.newBuilder()
+               .expireAfterAccess(state.getPropAsInt(GobblinMCEWriter.CACHE_EXPIRING_TIME,
+                   GobblinMCEWriter.DEFAULT_CACHE_EXPIRING_TIME), TimeUnit.HOURS)
+               .build());
+           HiveSpec existedSpec = hiveSpecCache.getIfPresent(partitionValue);
+           if (existedSpec != null) {
+             //if existedSpec is not null, it means we already registered this partition, so check whether we need to update the table/partition
+             if (gmce.getSchemaSource() == SchemaSource.SCHEMAREGISTRY) {
+               // Only in this case, we need to update the schema to the latest schema in schemaRegistry
+               schemaUpdateHelper(spec, topicName);
+             }
+             if (this.hiveRegister.needToUpdateTable(existedSpec.getTable(), spec.getTable())) {
+                registerSpec(tableKey, partitionValue, spec, hiveSpecCache);
+                return;
+             }
+             if (spec.getPartition().isPresent() &&
```

Comment:
so if condition in line and the one in line 165 both satisfied, `registerSpec(tableKey, partitionValue, spec, hiveSpecCache);` will be called twice?

Reply:
on line 167, it will directly return the method

Reply:
OK I missed that. Then why you need two branches?